proposals/xxxx-mesh-nodes.md

```diff
@@ -0,0 +1,329 @@
+Like a broadcast node input record, this must have an `SV_DispatchGrid` field
+indicating the grid size when `[NodeMaxDispatchGrid(x,y,z)]` is used.
+
+The same system value inputs supported for broadcast launch nodes and mesh
+shaders are supported in the same way here:
+`SV_DispatchThreadID`, `SV_GroupThreadID`, `SV_GroupIndex`, `SV_GroupID`,
+with the exception (for now) of `SV_ViewID`, which is not supported for this
+experimental feature.
+
+Outputs to other nodes are not allowed - the mesh node must be a leaf in the
+work graph.
+
+The new `mesh` node supports outputs inherited from the mesh shaders:
+`vertices`, `primitives`, and `indices`.  As with mesh shaders, the
+`SetMeshOutputCounts` intrinsic must be used before writing to these outputs.
+As with mesh shaders, the `vertices` and `primitives` outputs are combined to
+provide the inputs the pixel shader requires.
+Each of these require a user defined structure using semantics on fields, where
+fields of HLSL basic types will be used to generate signature elements in the
+same way they are for mesh shader signature handling.
+
+Amplification shaders are not supported in this context, since work graph node
+shaders provide the function of generating the mesh node dispatches on the GPU.
+
+Mesh nodes can also be used as one or more entrypoints in a work graph
+(standalone since they can't output to other nodes),
+in which case the input record (payload) is supplied from DispatchGraph()
+at the API just like any other work graph entrypoint input.
+
+A complete mesh node is defined at the API by first creating a
+[generic program](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#d3d12_generic_program_desc)
+including the mesh node function name, a pixel shader, and other optional state,
+then referring to that program in a
+[work graph program node](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#d3d12_program_node),
+which may optionally override node properties, such as the node ID.
+
+See the [DispatchMesh Launch Nodes](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#dispatchmesh-launch-nodes)
+section of the work graphs spec, along with sections linked from there, to
+complete the picture.
+
+In the future, we may add subobjects to specify more of these details within
+the DXIL library for convenience.
+
+### Example Syntax
+
+```cpp
+struct MyMeshRecord {
+  float3 perMeshConstant;
+  uint3 dispatchGrid : SV_DispatchGrid;
+}
+
+groupshared uint numVerts;
+groupshared uint numPrims;
+
+[shader("node")]
+[NodeLaunch("mesh")]
+[NodeIsProgramEntry]
+[NodeID("meshNode", 2)]
+[NodeLocalRootArgumentsTableIndex(4)]
+[NumThreads(4,4,1)]
+[NodeMaxDispatchGrid(16,16,1)]
+[OutputTopology("triangle")]
+void MyMeshNode(
+    DispatchNodeInputRecord<MyMeshRecord> nodeInput,
+    out vertices PerVertexValues verts[16],
+    out primitives PerPrimitiveValues prim[8],
+    out indices uint3 tris[16] )
+{
```

Comment:
Is it OK to call a normal mesh shader entry inside a mesh node to share code?

Reply:
I don't think that would be a good idea, even if the compiler allows you to do so.  Technically, it might work out (by accident) with inlining/optimization/copy elimination, but there's nothing in the language to preserve special argument handling to avoid copying everything between shared memory record/arrays and parameters that will be considered thread-local for the called mesh shader entry.  We could adjust internals to better support that scenario, but I don't think that's a priority at this time.  References (and appropriately assigning a different address space to params) could make a shared function called by both entry function types work properly in the future.